analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,233 @@
+                #    TCHE = ptcs_IP_signif_sorted[TCHE_index]
+                #    TCHE_x = ptcs_x_at_time_0[ptcs_IP_signif.index(TCHE)]
+                #    TCHE_y = ptcs_y_at_time_0[ptcs_IP_signif.index(TCHE)]
+                #if TCHE_x**2 + TCHE_y**2 - (0.05)**2 > 0:
+                #    TCHE = -99
+                #    TCHP = -99
+                #    TCHE_IP = -99
+                #    TCHP_IP = -99
+                #    TCHE_x = -99
+                #    TCHE_y = -99
+                #   TCHE_z = -99
+                #else:
+                #    TCHE_IP = ptcs_IP[ptcs_IP_signif.index(TCHE)]
+                #    TCHE_z = ptcs_z_at_time_0[ptcs_IP_signif.index(TCHE)]
+                #    TCHP = ptcs_IP_signif_sorted[TCHE_index + 1]
+                #    TCHP_IP = ptcs_IP[ptcs_IP_signif.index(TCHP)]
+            #if TCHE > 30 and event.K0s == 0 and event.Kp == 0 and event.L0 == 0 and event.S0 == 0 and event.Sp == 0 and event.Sm == 0:
+            #    import pdb; pdb.set_trace()
+            
+            jet.tags['IP_b_LL'] = IP_b_LL  if self.tag_IP_b_LL  else None
+            jet.tags['IPs_b_LL']= IPs_b_LL if self.tag_IPs_b_LL else None
+            jet.tags['TCHE'] = TCHE
+            jet.tags['TCHP'] = TCHP
+            jet.tags['TCHE_IP'] = TCHE_IP
+            jet.tags['TCHP_IP'] = TCHP_IP
+            jet.tags['TCHE_x'] = TCHE_x
+            jet.tags['TCHE_y'] = TCHE_y
+            jet.tags['TCHE_z'] = TCHE_z
+            jet.tags['TCHE_xy'] = (TCHE_x**2+TCHE_y**2)**0.5
+            jet.tags['TCHE_pt'] = TCHE_pt
+            jet.tags['TCHE_dr'] = TCHE_dr
+            
+            if hasattr(event, 'K0s') == True :
+                jet.tags['K0s'] = event.K0s
+            else :
+                jet.tags['K0s'] = -99
+            if hasattr(event, 'Kp') == True :
+                jet.tags['Kp'] = event.Kp
+            else :
+                jet.tags['Kp'] = -99
+            if hasattr(event, 'L0') == True :
+                jet.tags['L0'] = event.L0
+            else :
+                jet.tags['L0'] = -99
+            if hasattr(event, 'S0') == True :
+                jet.tags['S0'] = event.S0
+            else :
+                jet.tags['S0'] = -99
+            if hasattr(event, 'Sp') == True :
+                jet.tags['Sp'] = event.Sp
+            else :
+                jet.tags['Sp'] = -99
+            if hasattr(event, 'Sm') == True :
+                jet.tags['Sm'] = event.Sm
+            else :
+                jet.tags['Sm'] = -99
+            if hasattr(event, 'Muons') == True :
+                jet.tags['Muons'] = event.Muons
+            else :
+                jet.tags['Muons'] = -99
+        
         
+    def beginLoop(self, setup):
+        super(ImpactParameter, self).beginLoop(setup)
+        if hasattr(self.cfg_ana, 'num_IP') == False :
+            self.tag_IP_b_LL = False
+        else :
+            if hasattr(self.cfg_ana, 'denom_IP') == False :
+                self.tag_IP_b_LL = False
```

Comment:
why no exception? For next update of the PR, please specifically answer the comments you did not take into account, explaining why. (a bit difficult for me to keep track otherwise!)